wallet-new/src/Cardano/Wallet/API/V1/Types.hs

```diff
@@ -849,7 +852,50 @@ instance Arbitrary WalletAddress where
                               <*> arbitrary
                               <*> arbitrary
 
-type AccountIndex = Word32
+newtype AccountIndex = AccountIndex { getAccIndex :: Word32 }
+    deriving (Show, Eq, Ord, Generic)
+
+instance Bounded AccountIndex where
+    -- NOTE: minimum for hardened key. See https://iohk.myjetbrains.com/youtrack/issue/CO-309
+    minBound = AccountIndex 2147483648
+    maxBound = AccountIndex maxBound
+
+mkAccountIndex :: Word32 -> Either Text AccountIndex
+mkAccountIndex index | index >= getAccIndex minBound = Right $ AccountIndex index
+                     | otherwise = Left $ sformat
+                            ("mkAccountIndex: Account index should be in range ["%int%".."%int%"]")
```

Comment:
do we think it make sense to add additional constructor to `WalletError` which would capture this failure?

There are other places in this module where I can do this refactoring (possibly as part of https://iohk.myjetbrains.com/youtrack/issue/CBR-26 )

Reply:
Oui :) 
IMO, we should always favor explicit and typed error when we can. Especially when the error is a well-identified scenario that is reachable.